### PR TITLE
Add a refactored ZoneGetter package and filter nodes with providerID

### PR DESCRIPTION
**Pull request by songrx1997**
/assign @swetharepakula @bowei

### CODE REVIEW

File: pkg/zonegetter/zone_getter.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+// ZoneLister is a interface to manages lookups for GCE instances to zones.
+type ZoneLister interface {
+	ListZones(mode Mode) ([]string, error)
+	GetZoneForNode(name string) (string, error)
+}
+
+// ZoneGetter implements ZoneLister Interface and manages lookups for
+// GCE instances to zones.
+type ZoneGetter struct {
+	NodeInformer cache.SharedIndexInformer
```
Comment (bowei) [inline pkg/zonegetter/zone_getter.go:54]: this should be package private

File: pkg/zonegetter/zone_getter.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// GetZoneForNode returns the zone for a given node by looking up providerID.
+func (z *ZoneGetter) GetZoneForNode(name string) (string, error) {
```
Comment (bowei) [inline pkg/zonegetter/zone_getter.go:58]: Get is redundant

ZoneForNode

File: pkg/zonegetter/zone_getter.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	zone, err := getZone(node)
+	if err != nil {
+		return "", err
+	}
+	return zone, nil
+}
+
+// ListZones returns a list of zones containing nodes that satisfy the given
+// node filtering mode.
+func (z *ZoneGetter) ListZones(mode Mode) ([]string, error) {
```
Comment (bowei) [inline pkg/zonegetter/zone_getter.go:76]: ListZones => List

File: pkg/zonegetter/zone_getter.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+	return filtered, nil
+}
+
+func getZone(node *api_v1.Node) (string, error) {
+	zone, err := getZoneByProviderID(node.Spec.ProviderID)
+	if err != nil {
+		klog.Errorf("Failed to get zone information from node %q: %v", node.Name, err)
+		return "", err
+	}
```
Comment (bowei) [inline pkg/zonegetter/zone_getter.go:125]: if zone == "" { return "", fmt.Errorf("node %s has an empty zone", node.Name) }

File: pkg/zonegetter/zone_getter.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	return zone, nil
+}
+
+// getZoneByProviderID gets zone information from node provider id.
+// A providerID is build out of '${ProviderName}://${project-id}/${zone}/${instance-name}'
+func getZoneByProviderID(providerID string) (string, error) {
+	var providerIDRE = regexp.MustCompile(`^` + "gce" + `://([^/]+)/([^/]+)/([^/]+)$`)
+	matches := providerIDRE.FindStringSubmatch(providerID)
+	if len(matches) != 4 {
+		return "", fmt.Errorf("%w: providerID %q is not in valid format", ErrSplitProviderID, providerID)
```
Comment (bowei) [inline pkg/zonegetter/zone_getter.go:135]: providerID %q is not valid

File: pkg/zonegetter/zone_getter.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+		klog.Errorf("Failed to get zone information from node %q: %v", node.Name, err)
+		return "", err
+	}
+	return zone, nil
+}
+
+// getZoneByProviderID gets zone information from node provider id.
+// A providerID is build out of '${ProviderName}://${project-id}/${zone}/${instance-name}'
+func getZoneByProviderID(providerID string) (string, error) {
+	var providerIDRE = regexp.MustCompile(`^` + "gce" + `://([^/]+)/([^/]+)/([^/]+)$`)
```
Comment (bowei) [inline pkg/zonegetter/zone_getter.go:132]: you should move this outside at the package level rather than MustCompile everytime this func is called

File: pkg/zonegetter/zone_getter.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+		if predicate(nodes[i]) && nodes[i].Spec.ProviderID != "" {
+			filtered = append(filtered, nodes[i])
+		}
+	}
+
+	return filtered, nil
+}
+
+func getZone(node *api_v1.Node) (string, error) {
+	zone, err := getZoneByProviderID(node.Spec.ProviderID)
```
Comment (bowei) [inline pkg/zonegetter/zone_getter.go:121]: seems a little redundant to have getZone and getZoneByProviderID, can we just merge

File: pkg/zonegetter/zone_getter.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+func listNodesWithMode(nodeLister listers.NodeLister, mode Mode) ([]*api_v1.Node, error) {
+	var predicate utils.NodeConditionPredicate
+	switch mode {
+	case AllNodesMode:
+		predicate = utils.AllNodesPredicate
+	case CandidateNodesMode:
+		predicate = utils.CandidateNodesPredicate
+	case CandidateNodeIncludeUnreadyExcludeUpgradingMode:
+		predicate = utils.CandidateNodesPredicateIncludeUnreadyExcludeUpgradingNodes
+	}
```
Comment (bowei) [inline pkg/zonegetter/zone_getter.go:102]: default:
  return nil, <err>

File: pkg/zonegetter/zone_getter.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+	nodes, err := nodeLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	var filtered []*api_v1.Node
+	for i := range nodes {
+		// Filter nodes with predicate and exclude nodes without providerID
+		if predicate(nodes[i]) && nodes[i].Spec.ProviderID != "" {
```
Comment (bowei) [inline pkg/zonegetter/zone_getter.go:112]: We should put the state machine in a doc.go or markdown file in this package describing when provider ID will be ""

File: pkg/zonegetter/zone_getter.go (hunk 1 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go]: Done

File: pkg/zonegetter/zone_getter.go (hunk 3 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go]: Done

File: pkg/zonegetter/zone_getter.go (hunk 5 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go]: Done

File: pkg/zonegetter/zone_getter.go (hunk 6 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go]: Done

File: pkg/zonegetter/zone_getter.go (hunk 2 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go]: Done

File: pkg/zonegetter/zone_getter.go (hunk 4 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go:134]: Done

File: pkg/zonegetter/zone_getter.go (hunk 7 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go]: Done

File: pkg/zonegetter/zone_getter_test.go [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+			mode:      AllNodesMode,
+			expectLen: 2,
+		},
+		{
+			desc:      "List with CandidateNodesPredicate",
+			mode:      CandidateNodesMode,
+			expectLen: 1,
+		},
+		{
+			desc:      "List with CandidateNodesPredicateIncludeUnreadyExcludeUpgradingNodes",
```
Comment (swetharepakula) [inline pkg/zonegetter/zone_getter_test.go:121]: should we add nodes that are upgrading?

File: pkg/zonegetter/zone_getter_test.go [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+
+import (
+	"errors"
+	"testing"
+
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestList(t *testing.T) {
```
Comment (swetharepakula) [inline pkg/zonegetter/zone_getter_test.go:27]: can we add a test case that shows that we handle empty zones properly in list?

File: pkg/zonegetter/node-state.md [hunk 12]
```diff
@@ -0,0 +1,5 @@
+| Node State     | ProviderID | Consider it valid? |
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/zonegetter/node-state.md]: Done

File: pkg/zonegetter/zone_getter.go [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+// ZoneGetter implements ZoneLister Interface and manages lookups for
+// GCE instances to zones.
+type ZoneGetter struct {
+	nodeInformer cache.SharedIndexInformer
+}
+
+// ZoneForNode returns the zone for a given node by looking up providerID.
+func (z *ZoneGetter) ZoneForNode(name string) (string, error) {
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go]: Done

File: pkg/zonegetter/zone_getter.go [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+		}
+	}
+	return zones.List(), nil
+}
+
+// ListWithPredicate gets nodes that matches node filter mode.
+func listNodesWithMode(nodeLister listers.NodeLister, mode Mode) ([]*api_v1.Node, error) {
+	var filtered []*api_v1.Node
+
+	var predicate utils.NodeConditionPredicate
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go:108]: Done

File: pkg/zonegetter/zone_getter_test.go (hunk 10 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter_test.go:158]: Done

File: pkg/zonegetter/zone_getter_test.go (hunk 11 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter_test.go:28]: Done

File: pkg/zonegetter/zone_getter.go [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+)
+
+var ErrProviderIDNotFound = errors.New("providerID not found")
+var ErrSplitProviderID = errors.New("error splitting providerID")
+var ErrNodeNotFound = errors.New("node not found")
+
+// providerIDRE is the regex to process providerID.
+var providerIDRE = regexp.MustCompile(`^` + "gce" + `://([^/]+)/([^/]+)/([^/]+)$`)
+
+// ZoneLister is a interface to manages lookups for GCE instances to zones.
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go:52]: We also have a lot of tests depend on the [FakeZoneGetter](https://github.com/kubernetes/ingress-gce/blob/master/pkg/neg/types/fakes.go#L52). If we remove the interface, we will need to implement the FakeZoneGetter in a different way. What do you think?

File: pkg/zonegetter/zone_getter.go [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+
+// NewZoneGetter initialize a ZoneGetter in normal mode.
+func NewZoneGetter(nodeInformer cache.SharedIndexInformer) *ZoneGetter {
+	return &ZoneGetter{
+		mode:         Normal,
+		nodeInformer: nodeInformer,
+	}
+}
+
+// NewZoneGetter initialize a fake ZoneGetter in normal mode.
```
Comment (swetharepakula) [inline pkg/zonegetter/zone_getter.go:186]: move this into the test file

File: pkg/zonegetter/zone_getter_test.go [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+		},
+		{
+			desc:      "List with CandidateNodesPredicateIncludeUnreadyExcludeUpgradingNodes",
+			filter:    CandidateNodeIncludeUnreadyExcludeUpgradingFilter,
+			expectLen: 2,
+		},
+	} {
+		zones, _ := zoneGetter.List(tc.filter, klog.TODO())
+		if len(zones) != tc.expectLen {
+			t.Errorf("For test case %q, got %d zones, want %d,", tc.desc, len(zones), tc.expectLen)
```
Comment (swetharepakula) [inline pkg/zonegetter/zone_getter_test.go:166]: can we add an expectation that no empty zones are in the list

File: pkg/zonegetter/zone_getter_test.go [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+	"errors"
+	"reflect"
+	"testing"
+
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+)
+
+func TestList(t *testing.T) {
```
Comment (swetharepakula) [inline pkg/zonegetter/zone_getter_test.go:29]: we need tests for the simple zone getter mode as well

Comment (swetharepakula) [reply pkg/zonegetter/zone_getter_test.go:29]: thought i deleted the comment. There is a different simple zone getter test which covers it.

File: pkg/zonegetter/zone_getter.go [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+		predicate = utils.CandidateNodesPredicate
+	case CandidateNodeIncludeUnreadyExcludeUpgradingFilter:
+		predicate = utils.CandidateNodesPredicateIncludeUnreadyExcludeUpgradingNodes
+	default:
+		return filtered, nil
+	}
+
+	nodes, err := nodeLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go:141]: The err will be logged in the caller List(). Since this function is a internal helper function, it will be duplicate if we log it twice.

File: pkg/zonegetter/zone_getter.go [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+
+type Mode int
+type Filter string
+
+const (
+	Normal Mode = iota
+	Simple
+)
+
+const (
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go:44]: If we make these enums, it will be unclear which filter we are using in the logs.

File: pkg/zonegetter/zone_getter.go [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+	AllNodesFilter                                    = "AllNodesFilter"
+	CandidateNodesFilter                              = "CandidateNodesFilter"
+	CandidateNodeIncludeUnreadyExcludeUpgradingFilter = "CandidateNodeIncludeUnreadyExcludeUpgradingFilter"
+)
+
+var ErrProviderIDNotFound = errors.New("providerID not found")
+var ErrSplitProviderID = errors.New("error splitting providerID")
+var ErrNodeNotFound = errors.New("node not found")
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go:52]: My plan is handle this outside like this https://github.com/kubernetes/ingress-gce/commit/5bf9c82d0f6900ccf38a3835219ef685daa39a9a#diff-f64364884edd3d19cda733b2ca7bda358bfba8fca128782330fe53d1fde6f08aL371-R385

We can classify errors for metrics outside with our current metrics.

File: pkg/zonegetter/zone_getter.go [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+	nodes, err := nodeLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range nodes {
+		// Filter nodes with predicate and exclude nodes without providerID
+		if predicate(nodes[i]) {
+			filtered = append(filtered, nodes[i])
+		}
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go:148]: Done

File: pkg/zonegetter/zone_getter.go (hunk 16 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go]: Done

File: pkg/zonegetter/zone_getter_test.go (hunk 17 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter_test.go:170]: Done

File: pkg/zonegetter/node-state.md [hunk 23]
```diff
@@ -0,0 +1,5 @@
+| Node State     | ProviderID | Controller Processes/Uses the Node |
```
[the remark this answers is not in the document]
Comment (songrx1997) [reply pkg/zonegetter/node-state.md]: Done

File: pkg/zonegetter/zone_getter.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (bowei) [reply pkg/zonegetter/zone_getter.go:42]: you can always add the String() method for logging purposes, but usually it's ok to keep as string const to skip the step

```
type enumType int

func (e enumType) String() string {
  switch e {
  case ...: return "enumType1"
  }
  return "invalid enumType"
}

```

File: pkg/zonegetter/zone_getter.go [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/ingress-gce/pkg/utils"
+	"k8s.io/klog/v2"
+)
+
+type Mode int
+type Filter string
+
+const (
+	Normal Mode = iota
```
Comment (bowei) [inline pkg/zonegetter/zone_getter.go:38]: This needs docstring to explain what these mean

Comment (bowei) [reply pkg/zonegetter/zone_getter.go:38]: Also, we should probably change the names as Normal and Simple are almost meaningless.

File: pkg/zonegetter/zone_getter.go [hunk 25]
```diff
@@ -0,0 +1,184 @@
...
+	Simple
+)
+
+const (
+	AllNodesFilter                                    = "AllNodesFilter"
+	CandidateNodesFilter                              = "CandidateNodesFilter"
+	CandidateNodeIncludeUnreadyExcludeUpgradingFilter = "CandidateNodeIncludeUnreadyExcludeUpgradingFilter"
```
Comment (bowei) [inline pkg/zonegetter/zone_getter.go:45]: is there not a better name?

File: pkg/zonegetter/zone_getter.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/zonegetter/zone_getter.go:139]: for now, since there is only one caller, that's okay. in the future it might not be so. But we can add it then

File: pkg/zonegetter/zone_getter.go (hunk 24 above)
Comment (swetharepakula) [reply pkg/zonegetter/zone_getter.go:38]: Perhaps SingleZone, or since there is only one usecase for this (Non-GCP). We could have a Non-GCP and GCP modes?

Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go]: Done

File: pkg/zonegetter/zone_getter.go (hunk 25 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go]: Renamed it since `candidateNodes` has already excluded upgrading nodes.

File: pkg/zonegetter/zone_getter.go [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+		return nil, err
+	}
+
+	for _, node := range nodes {
+		// Filter nodes with predicate and exclude nodes without providerID
+		if predicate(node) {
+			filtered = append(filtered, node)
+		} else if len(nodes) <= 50 {
+			logger.Info("Filtered out node when listing node zones", "node", node.Name, "filter", filter)
+		}
```
Comment (swetharepakula) [inline pkg/zonegetter/zone_getter.go:151]: perhaps we do a single line after the loop instead?

File: pkg/zonegetter/zone_getter.go [hunk 27]
```diff
@@ -0,0 +1,187 @@
...
+	nodeLister := z.nodeInformer.GetIndexer()
+	zones := sets.String{}
+	nodes, err := listNodesWithFilter(listers.NewNodeLister(nodeLister), filter, logger)
+	if err != nil {
+		logger.Error(err, "Failed to list nodes")
+		return zones.List(), err
+	}
+	for _, n := range nodes {
+		zone, err := getZone(n)
+		if err != nil {
```
Comment (swetharepakula) [inline pkg/zonegetter/zone_getter.go:115]: should this check for the empty zone error?

Comment (swetharepakula) [reply pkg/zonegetter/zone_getter.go:115]: read the code wrong, ignore the statement

**Review by swetharepakula — COMMENTED** [review]
can you move the zone getter package to the pkg/utils.

We should reduce the number of high level packages.

File: pkg/utils/zonegetter/zone_getter_test.go [hunk 28]
```diff
@@ -0,0 +1,348 @@
...
+			filter:    AllNodesFilter,
+			expectLen: 3,
+		},
+		{
+			desc:      "List with CandidateNodesPredicate",
+			filter:    CandidateNodesFilter,
+			expectLen: 1,
+		},
+		{
+			desc:      "List with CandidateNodesPredicateIncludeUnreadyExcludeUpgradingNodes",
```
Comment (swetharepakula) [inline pkg/utils/zonegetter/zone_getter_test.go:163]: nit: update the predicate with the new name

File: pkg/zonegetter/zone_getter.go (hunk 26 above)
Comment (songrx1997) [reply pkg/zonegetter/zone_getter.go]: Done

File: pkg/utils/zonegetter/zone_getter_test.go (hunk 28 above)
Comment (songrx1997) [reply pkg/utils/zonegetter/zone_getter_test.go]: Done

**Review by swetharepakula — COMMENTED** [review]
/lgtm
/approve
